### PR TITLE
[X86] Fast AVX-512-VNNI vpdpwssd tuning

### LINKED ISSUE #84182: `vpdpwssd` instruction not generated despite giving better performance than `vpmaddwd+vpaddd` expansion
Filing this Issue with two Compiler Explorer testcases - one in LLVM IR and the other in C.

# LLVM IR testcase

Compiler Explorer link: https://godbolt.org/z/3Wf1cfEo1

Problem: when the parent function has `"target-cpu"="znver4"`, the `@llvm.x86.avx512.vpdpwssd.512` intrinsic fails to compile to the expected AVX-512-VNNI `vpdpwssd` instruction and instead generates a (`vpmaddwd`, `vpaddd`) fallback implementation.

The problem only reproduces with `"target-cpu"="znver4"` and not with `"target-cpu"="cascadelake"` or when `"target-cpu"` is simply omitted, as shown in the Compiler Explorer link.

Inlining the LLVM IR testcase here for completeness:

```llvm
; Testing with "target-cpu"="znver4" in attributes #0 below. Nothing else changes between testcases.

define dso_local <8 x i64> @foo(<8 x i64> noundef %0, <8 x i64> noundef %1, <8 x i64> noundef %2) local_unnamed_addr #0 {
  %4 = bitcast <8 x i64> %0 to <16 x i32>
  %5 = bitcast <8 x i64> %1 to <16 x i32>
  %6 = bitcast <8 x i64> %2 to <16 x i32>
  %7 = tail call <16 x i32> @llvm.x86.avx512.vpdpwssd.512(<16 x i32> %4, <16 x i32> %5, <16 x i32> %6)
  %8 = bitcast <16 x i32> %7 to <8 x i64>
  ret <8 x i64> %8
}

declare <16 x i32> @llvm.x86.avx512.vpdpwssd.512(<16 x i32>, <16 x i32>, <16 x i32>) #1

declare void @llvm.dbg.value(metadata, metadata, metadata) #2

attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "min-legal-vector-width"="512" "no-trapping-math"="true" "stack-protector-buffer-size"="8"
  "target-cpu"="znver4" "target-features"="+adx,+aes,+avx,+avx2,+avx512bf16,+avx512bitalg,+avx512bw,+avx512cd,+avx512dq,+avx512f,+avx512ifma,+avx512vbmi,+avx512vbmi2,+avx512vl,+avx512vnni,+avx512vpopcntdq,+bmi,+bmi2,+clflushopt,+clwb,+clzero,+crc32,+cx16,+cx8,+evex512,+f16c,+fma,+fsgsbase,+fxsr,+gfni,+invpcid,+lzcnt,+mmx,+movbe,+mwaitx,+pclmul,+pku,+popcnt,+prfchw,+rdpid,+rdpru,+rdrnd,+rdseed,+sahf,+sha,+shstk,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+sse4a,+ssse3,+vaes,+vpclmulqdq,+wbnoinvd,+x87,+xsave,+xsavec,+xsaveopt,+xsaves" }
attributes #1 = { mustprogress nocallback nofree nosync nounwind willreturn memory(none) }
attributes #2 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

# C testcase

Compiler Explorer link: https://godbolt.org/z/z6xxf35hG

Problem: when compiling with `-march=znver4`, the `_mm512_dpwssd_epi32` intrinsic fails to compile to the expected AVX-512-VNNI `vpdpwssd` instruction and instead generates a (`vpmaddwd`, `vpaddd`) fallback implementation.

The problem only reproduces with `-march=znver4` and not with `-march=cascade` or when `-march` is simply omitted and individual AVX-512 features are passed instead, as shown in the Compiler Explorer link.

Inlining the C testcase here for completeness:

```c
#include <immintrin.h>

__m512i foo(__m512i x, __m512i y, __m512i z) {
    return _mm512_dpwssd_epi32(x, y, z);
}
```

### PR DESCRIPTION
**Pull request by ganeshgit**
Adding a tuning feature to resolve #84182. Generation of vpdpwssd (instead of vpmaddwd + vpaddd sequence) will be retained when the tuning is enabled.

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -48215,7 +48215,7 @@ static SDValue combineAndShuffleNot(SDNode *N, SelectionDAG &DAG,
 
   // We do not split for SSE at all, but we need to split vectors for AVX1 and
   // AVX2.
-  if (!Subtarget.useAVX512Regs() && VT.is512BitVector() && 
+  if (!Subtarget.useAVX512Regs() && VT.is512BitVector() &&
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:48218]: Unrelated change

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 2]
```diff
@@ -10572,18 +10572,23 @@ bool X86InstrInfo::getMachineCombinerPatterns(
   case X86::VPDPWSSDrm:
   case X86::VPDPWSSDYrr:
   case X86::VPDPWSSDYrm: {
-    Patterns.push_back(MachineCombinerPattern::DPWSSD);
-    return true;
+    if (!Subtarget.hasFastPWSSD()) {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:10575]: Put this at the beginning?
```
if (!Subtarget.hasFastPWSSD())
  return TargetInstrInfo::getMachineCombinerPatterns(Root, Patterns,
                                                     DoRegPressureReduce);
```

File: llvm/lib/Target/X86/X86.td [hunk 3]
```diff
@@ -683,6 +683,12 @@ def TuningFastGather
     : SubtargetFeature<"fast-gather", "HasFastGather", "true",
                        "Indicates if gather is reasonably fast (this is true for Skylake client and all AVX-512 CPUs)">;
 
+// Generate vpdpwssd instead of vpmaddwd+vpaddd sequence.
+def TuningFastPWSSD
+    : SubtargetFeature<
+          "fast-pwssd", "HasFastPWSSD", "true",
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86.td:689]: Shouldn't it be fast-dpwssd?

File: llvm/test/CodeGen/X86/vpdpwssd.ll [hunk 4]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mcpu=znver4 | FileCheck %s
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/vpdpwssd.ll:2]: The NOTE should be the first line to keep regeneration scripts happy

File: llvm/test/CodeGen/X86/vpdpwssd.ll [hunk 5]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown -mcpu=znver4 | FileCheck %s
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/vpdpwssd.ll:1]: Add an extra RUN that takes the -mattr=+avx512vnni,+fast-dpwssd

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 2 above)
Comment (RKSimon) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:10575]: Can you remove the default, break if we don't use the MachineCombinerPattern::DPWSSD pattern and move this to after the switch:
```
return TargetInstrInfo::getMachineCombinerPatterns(Root, Patterns, DoRegPressureReduce);
```

**Review by RKSimon — CHANGES_REQUESTED** [review]
This patch needs to be cleanly rebased on trunk (force push is OK in PR branchs)

**Comment by ganeshgit** [comment]
> This patch needs to be cleanly rebased on trunk (force push is OK in PR branchs)
I think my overnight rebase scripts screwed the fixups. I will close this and I will raise a new pull request.
